astropy/io/misc/hdf5.py

```diff
@@ -266,11 +288,20 @@ def write_table_hdf5(table, output, path=None, compression=False,
     if serialize_meta:
         header_yaml = meta.get_yaml_from_table(table)
 
-        try:
-            dset.attrs[META_KEY] = [h.encode('utf8') for h in header_yaml]
-        except Exception as e:
-            warnings.warn("Attributes could not be written to the output HDF5 "
-                          "file: {0}".format(e))
+        header_encoded = [h.encode('utf-8') for h in header_yaml]
+        if compatibility_mode:
+            warnings.warn("compatibility mode for writing is deprecated",
+                          AstropyDeprecationWarning)
+            try:
+                dset.attrs[META_KEY] = header_encoded
+            except Exception as e:
+                warnings.warn(
+                "Attributes could not be written to the output HDF5 "
+                "file: {0}".format(e))
+
+        else:
+            dset2 = output_group.create_dataset(meta_path(name),
```

Comment:
Another low priority fix, if you not using `dset2` then no real need to define it.  The return value of `output_group...` can just be dropped on the floor.